apps/service_profile/lib/profile/feed.ex

```diff
@@ -0,0 +1,98 @@
+defmodule Profile.Feed do
+  use Supervisor
+
+  @type init_opts :: [
+          name: GenServer.name(),
+          extract: Extract.t()
+        ]
+
+  @spec start_link(init_opts) :: GenServer.on_start()
+  def start_link(opts) do
+    server_opts = Keyword.take(opts, [:name])
+    Supervisor.start_link(__MODULE__, opts, server_opts)
+  end
+
+  @impl true
+  def init(opts) do
+    extract = Keyword.fetch!(opts, :extract)
+
+    children = [
+      {Profile.Feed.Flow,
+       dataset_id: extract.dataset_id,
+       subset_id: extract.subset_id,
+       from_specs: [
+         {Profile.Feed.Producer, extract: extract}
+       ],
+       into_specs: [
+         {Profile.Feed.Consumer, dataset_id: extract.dataset_id, subset_id: extract.subset_id}
+       ],
+       reducers: determine_reducers(extract.dictionary, [], [Access.key(:value)])}
+    ]
+
+    Supervisor.init(children, strategy: :one_for_one)
+  end
+
+  def determine_reducers([], reducers), do: reducers
+
+  def determine_reducers([{_, _} | _] = dictionaries, reducers) do
+    Enum.reduce(dictionaries, reducers, fn {dictionary, path}, acc ->
+      determine_reducers(dictionary, acc, path)
+    end)
+  end
+
+  def determine_reducers(dictionary, reducers, path) do
+    reducers =
+      find_temporal_range(reducers, dictionary, path)
+      |> find_bounding_box(dictionary, path)
+
+    dictionaries =
+      Enum.filter(dictionary, fn %type{} -> type == Dictionary.Type.Map end)
+      |> Enum.map(fn map -> {map.dictionary, path ++ [map.name]} end)
+
+    determine_reducers(dictionaries, reducers)
+  end
+
+  defp find_bounding_box(reducers, dictionary, path) do
+    longitude = Enum.find(dictionary, fn %type{} -> type == Dictionary.Type.Longitude end)
+    latitude = Enum.find(dictionary, fn %type{} -> type == Dictionary.Type.Latitude end)
+
+    case {longitude, latitude} do
+      {long, lat} when long != nil and lat != nil ->
+        reducer =
+          Profile.Reducer.BoundingBox.new(
+            longitude_path: path ++ [long.name],
+            latitude_path: path ++ [lat.name]
+          )
+
+        add_to_reducers(reducers, reducer)
+
+      _ ->
+        reducers
+    end
+  end
+
+  defp find_temporal_range(reducers, dictionary, path) do
+    case Enum.find(dictionary, fn %type{} -> type == Dictionary.Type.Timestamp end) do
+      nil ->
+        reducers
+
+      field ->
+        reducer = Profile.Reducer.TemporalRange.new(path: path ++ [field.name])
+        add_to_reducers(reducers, reducer)
+    end
+  end
+
+  defp add_to_reducers(reducers, reducer) do
+    case find_reducer(reducers, reducer) do
+      nil -> reducers ++ [reducer]
+      _ -> reducers
+    end
+  end
+
+  defp find_reducer(reducers, %struct{}) do
+    Enum.find(reducers, fn
+      %^struct{} -> true
+      _ -> false
+    end)
+  end
```

Comment:
I really struggled to understand the logic here. I re-read it a few times before it clicked, but I think the reason is because we don't always `add_to_reducers`. When a reducer exists, we don't add a new one, right? That was hard to decipher for me.

Reply:
I'm going to suggest this change:

```elixir
  defp find_temporal_range(reducers, dictionary, path) do
    case Enum.find(dictionary, fn %type{} -> type == Dictionary.Type.Timestamp end) do
      nil ->
        reducers

      field ->
        reducer = Profile.Reducer.TemporalRange.new(path: path ++ [field.name])

        case includes_reducer?(reducers, reducer) do
          true -> reducers
          false -> reducers ++ [reducer]
        end
    end
  end

  defp includes_reducer?(reducers, reducer) do
    case find_reducer(reducers, reducer) do
      nil -> false
      _ -> true
    end
  end

  defp find_reducer(reducers, %struct{}) do
    Enum.find(reducers, fn
      %^struct{} -> true
      _ -> false
    end)
  end
```

Reply:
IMO, by using a predicate like `includes_reducer?/2` up front, it's clearer that we're adding a reducer to the list when it doesn't already exist.

Reply:
is `includes_reducer?/2` really giving you anything in that scenario? You're basically just reversing the result of `find_reduce/2` then.

If that's the case, why not just:
```
unless find_reducer(reducers, reducer) do
  reducers ++ [reducer]
else
  reducers
end
```

Reply:
and then name `find_reducer/2` whatever you want to be clear if the name is still problematic for the reason you mentioned.

Reply:
That is fair but I would like to avoid the nested case statements

Reply:
my suggestion wouldn't use any case statements

Reply:
```elixir
defp find_temporal_range(reducers, dictionary, path) do
  case Enum.find(dictionary, fn %type{} -> type == Dictionary.Type.Timestamp end) do
    nil ->
      reducers
    field ->
      reducer = Profile.Reducer.TemporalRange.new(path: path ++ [field.name])

     unless find_reducer(reducers, reducer) do
       reducers ++ [reducer]
     else
       reducers
     end
  end
end

defp find_reducer(reducers, %struct{}) do
  Enum.find(reducers, fn
    %^struct{} -> true
    _ -> false
  end)
end
```
I take that back; it doesn't _add_ any case statements.

Reply:
I took a crack at rewriting it from scratch, let me know if it is better or worse.